CMakeLists.txt

```diff
@@ -16,8 +16,10 @@ endif ()
 ## Main library build
 ##
 
-set(CMAKE_CXX_VISIBILITY_PRESET hidden)
-set(CMAKE_VISIBILITY_INLINES_HIDDEN YES)
+if (BUILD_SHARED_LIBS)
+    set(CMAKE_CXX_VISIBILITY_PRESET hidden)
+    set(CMAKE_VISIBILITY_INLINES_HIDDEN YES)
+endif()

```

Comment:
On `CMAKE_CXX_VISIBILITY_PRESET`:

With CMake 3.3 (`CMP0063`), CMake explicitly decided to honor visibility properties also for static libraries, so that they can be linked into a shared library later:
https://cmake.org/cmake/help/latest/policy/CMP0063.html#policy:CMP0063
This seems to be an argument against guarding `CMAKE_CXX_VISIBILITY_PRESET hidden` by `BUILD_SHARED_LIBS`.

I also don't see how the guard actually contributes to the aim of this PR, "do not force export the symbols when building statically". Isn't this setting what is desired to achieve the aim, i.e hide symbols by default?

Reply:
When someone builds tinyxml2 itself as a shared library it is useful to hide the unnecessary symbols, but when they want to use it privately, tinyxml2 should not force its symbols to be exported from some other shared library. 

Again, this is consistent with the behaviour of famous libraries like spdlog

https://github.com/gabime/spdlog/blob/d546201f127c306ec8a0082d57562a05a049af77/include/spdlog/common.h#L23-L43

Reply:
The lines I commented here set the *default visibility* to  hidden. Is this wrong for static builds?

I didn't mention the changes to `TINYXML2_EXPORT` which control the visibility of symbols marked by `TINYXML2_LIB`.
But isn't this the full picture?

|  when | linkage | default visibility | `TINYXML2_LIB` vis. |
|----------|----------|----------|----------|
| before PR | shared | hidden | default |
| before PR | static | hidden | default |
| after PR | shared | hidden | default |
| after PR | static | default | default |

i.e. the net effect is making hidden symbols visible for static builds?

Reply:
The source override forces a symbol to be of some kind. When nothing is specified, the decision is deferred to the final build using the compiler flag. But when it is specified to be default, no matter what the symbols will be exported.

Reply:
AFAIU unspecified visibility prevents some optimizations, and [there are linker options for gcc and xcode](https://stackoverflow.com/questions/2222162/how-to-apply-fvisibility-option-to-symbols-in-static-libraries) to make symbols from a static lib excluded from a derived shared lib.

Reply:
The previous default prevented the inlining optimization for normal users that do not use such non-conventional flags. 
When the symbols are forced to be exported in the static library, they will be also exported in the shared library unless someone knows about `-Wl,--exclude-libs,ALL`. I haven't personally tested this flag, I don't know if it is supported on Windows, and if CMake knows about it.

But after this PR, those who do the last compilation can simply pass `-fvisiblity=hidden` or set `CMAKE_CXX_VISIBILITY_PRESET` to make everything hidden. No de-optimization happens.

Reply:
Okay, thanks. Resolved for me.